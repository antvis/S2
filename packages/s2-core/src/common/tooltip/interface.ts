import { SpreadSheet } from '../../sheet-type';
import { SortParam } from '../interface';

export type DataItem = Record<string, any>;

export interface IMenu {
  readonly id: string;
  readonly icon?: any;
  readonly text?: string;
  readonly children?: IMenu[]; // subMenu
}

export interface OperatorProps {
  readonly onClick: (...params) => void;
  readonly menus: IMenu[];
}

export interface Position {
  readonly x: number;
  readonly y: number;
  readonly tipHeight?: number;
}

export type ListItem = {
  name: string;
  value: string | number;
  icon?: string;
};

export interface SortQuery {
  [key: string]: string;
}

export interface TooltipOptions {
  hideSummary?: boolean;
  // button action on the top
  operator?: OperatorProps;
  enterable?: boolean;
  // totals or not
  isTotals?: boolean;
  showSingleTips?: boolean;
  [key: string]: any;
}

export interface SummaryProps {
  readonly name: string;
  readonly value: number | string;
  readonly selectedData: DataItem[];
}

export interface TipsProps {
  readonly tips: string;
}

export interface OperationProps {
  plot: SpreadSheet;
  sortFieldId: string;
  sortQuery: SortQuery;
}

export interface OperationState {
  sortParam: SortParam;
}

export type DetailProps = {
  list: ListItem[];
};

export type InterpretationProps = {
  name: string;
  icon?: any;
  text?: string;
  render?: React.ElementType;
};

export type InfosProps = {
  infos: string;
};

export type ShowProps = {
  position: Position;
  data?: DataProps;
  options?: TooltipOptions;
  element?: React.ReactElement;
};

export type DataProps = {
  summaries?: SummaryProps[];
  details?: ListItem[];
  headInfo?: HeadInfo;
  tips?: string;
  infos?: string;
  interpretation?: InterpretationProps;
};

export type HeadInfo = {
  rows: ListItem[];
  cols: ListItem[];
};
