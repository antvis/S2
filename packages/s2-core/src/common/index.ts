export * from './constant';
export * from './debug';
export * from './i18n';
export * from './icons';
export * from './interface';
export * from './store';
