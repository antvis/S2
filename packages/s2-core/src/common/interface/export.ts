import type { SpreadSheet } from '../../sheet-type';
import type { DataItem, CellMeta } from '..';
import { EXTRA_FIELD } from '..';

export type MatrixPlainTransformer = (
  data: DataItem[][],
  separator?: string,
) => CopyablePlain;

export type MatrixHTMLTransformer = (data: DataItem[][]) => CopyableHTML;

export enum CopyMIMEType {
  PLAIN = 'text/plain',
  HTML = 'text/html',
}

export type CopyableItem = {
  type: CopyMIMEType;
  content: string;
};

export type CopyablePlain = {
  type: CopyMIMEType.PLAIN;
  content: string;
};

export type CopyableHTML = {
  type: CopyMIMEType.HTML;
  content: string;
};

export type CopyableList = [CopyablePlain, CopyableHTML];

export type Copyable = CopyableItem | CopyableItem[];

export type FormatOptions =
  | boolean
  | {
      formatHeader?: boolean;
      formatData?: boolean;
    };

export interface Transformer {
  [CopyMIMEType.PLAIN]: MatrixPlainTransformer;
  [CopyMIMEType.HTML]: MatrixHTMLTransformer;
}

export interface CopyOrExportConfig {
  selectedCells?: CellMeta[];
  formatOptions?: FormatOptions;
  separator?: string;
  customTransformer?: (transformer: Transformer) => Partial<Transformer>;
  isAsyncExport?: boolean;
}

export interface CopyAndExportUnifyConfig {
  separator: string;
  formatHeader: boolean;
  formatData: boolean;
  selectedCells: CellMeta[];
  transformers: Transformer;
  isAsyncExport: boolean;
}

export interface CopyAllDataParams {
  sheetInstance: SpreadSheet;
  split?: string;
  formatOptions?: FormatOptions;
  customTransformer?: (transformer: Transformer) => Partial<Transformer>;
  /** 是否开启异步导出 */
  isAsyncExport?: boolean;
}

export interface SheetCopyConstructorParams {
  spreadsheet: SpreadSheet;
  config: CopyOrExportConfig;
  isExport?: boolean;
}

export type MeasureQuery =
  | { [EXTRA_FIELD]: string | undefined }
  // eslint-disable-next-line @typescript-eslint/ban-types
  | {};
