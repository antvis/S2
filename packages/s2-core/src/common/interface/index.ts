export * from './basic';
export * from './condition';
export * from './emitter';
export * from './events';
export * from './frame';
export * from './hooks';
export * from './interaction';
export * from './node';
export * from './resize';
export * from './s2DataConfig';
export * from './s2Options';
export * from './scroll';
export * from './store';
export * from './theme';
export * from './tooltip';
