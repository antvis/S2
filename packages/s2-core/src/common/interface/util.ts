export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
