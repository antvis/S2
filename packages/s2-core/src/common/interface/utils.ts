export type PickEssential<O> = {
  [K in keyof O as Pick<Partial<O>, K> extends Pick<O, K> ? never : K]: O[K];
};
