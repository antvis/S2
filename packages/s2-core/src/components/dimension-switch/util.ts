import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
