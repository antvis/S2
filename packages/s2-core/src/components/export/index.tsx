import React from 'react';
import { Menu, Dropdown, message } from 'antd';
import cx from 'classnames';
import { DotIcon } from '../icons/index';
import { copyData, copyToClipboard, download } from '../../utils/export';
import { SpreadSheet } from '@/sheet-type';
import { S2_PREFIX_CLS } from '@/common/constant/classnames';

export interface DataSet {
  icon?: React.ReactNode;
  name: string;
  value: string;
  type?: 'text' | 'location' | 'date';
  disabled?: boolean;
}

export interface ExportCfgProps {
  open: boolean;
  style?: React.CSSProperties;
  className?: string;
  icon?: React.ReactNode;
  copyOriginalText?: string;
  copyFormatText?: string;
  downloadOriginalText?: string;
  downloadFormatText?: string;
  successText?: string;
  errorText?: string;
  fileName?: string;
}
export interface ExportProps extends ExportCfgProps {
  sheet: SpreadSheet;
}

export const Export: React.FC<ExportProps> = ({
  className,
  icon,
  copyOriginalText = '复制原始数据',
  copyFormatText = '复制格式化数据',
  downloadOriginalText = '下载原始数据',
  downloadFormatText = '下载格式化数据',
  successText = '操作成功',
  errorText = '操作失败',
  sheet,
  fileName = 'sheet',
  ...restProps
}) => {
  const PRE_CLASS = `${S2_PREFIX_CLS}-export`;

  const exportData = (isFormat: boolean) => {
    const data = copyData(sheet, '\t', isFormat);
    if (copyToClipboard(data)) {
      message.success(successText);
    } else {
      message.error(errorText);
    }
  };

  const downLoadData = (isFormat: boolean) => {
    const data = copyData(sheet, ',', isFormat);
    try {
      download(data, fileName);
      message.success(successText);
    } catch (err) {
      message.error(errorText);
    }
  };

  const menu = (
    <Menu>
      <Menu.Item key="copyOriginal" onClick={() => exportData(false)}>
        <a href="`javascript:void(0)`">{copyOriginalText}</a>
      </Menu.Item>
      <Menu.Item key="copyFormat" onClick={() => exportData(true)}>
        <a href="`javascript:void(0)`">{copyFormatText}</a>
      </Menu.Item>
      <Menu.Item key="downloadOriginal" onClick={() => downLoadData(false)}>
        <a href="`javascript:void(0)`">{downloadOriginalText}</a>
      </Menu.Item>
      <Menu.Item key="downloadFormat" onClick={() => downLoadData(true)}>
        <a href="`javascript:void(0)`">{downloadFormatText}</a>
      </Menu.Item>
    </Menu>
  );

  return (
    <Dropdown
      overlay={menu}
      trigger={['click']}
      className={cx(PRE_CLASS, className)}
      {...restProps}
    >
      <a
        className="ant-dropdown-link"
        key="export"
        onClick={(e) => e.preventDefault()}
      >
        {icon || <DotIcon />}
      </a>
    </Dropdown>
  );
};
