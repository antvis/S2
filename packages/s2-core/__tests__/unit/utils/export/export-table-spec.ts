import { slice } from 'lodash';
import { data as originData } from 'tests/data/mock-dataset.json';
import { assembleDataCfg, assembleOptions } from '../../../util';
import { getContainer } from '../../../util/helpers';
import { TableSheet } from '@/sheet-type';
import { asyncGetAllPlainData } from '@/utils';
import { NewTab, NewLine } from '@/common';
import { CopyMIMEType } from '@/common/interface/export';

describe('TableSheet Export Test', () => {
  it('should export correct data with series number', async () => {
    const s2 = new TableSheet(
      getContainer(),
      assembleDataCfg({
        meta: [
          {
            field: 'type',
            name: '产品类型',
            formatter: (type) => `${type}产品`,
          },
        ],
        fields: {
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
        data: slice(originData, 0, 10),
      }),
      assembleOptions({
        seriesNumber: {
          enable: true,
        },
      }),
    );

    await s2.render();

    function testCase(data: string): void {
      const rows = data.split(NewLine);
      const headers = rows[0].split(NewTab);

      expect(slice(rows, 0, 5)).toMatchSnapshot();

      // 33行数据 包括一行列头
      expect(rows).toHaveLength(11);

      // 6列数据 包括序列号
      rows.forEach((e) => {
        expect(e.split(NewTab)).toHaveLength(6);
      });

      expect(headers).toEqual([
        '序号',
        'province',
        'city',
        '产品类型',
        'sub_type',
        'number',
      ]);
    }

    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: NewTab,
      formatOptions: {
        formatHeader: true,
      },
    });

    testCase(data);

    const asyncData = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: NewTab,
      formatOptions: {
        formatHeader: true,
      },
      isAsyncExport: true,
    });

    testCase(asyncData);
  });

  it('should export correct data with no series number', async () => {
    const s2 = new TableSheet(
      getContainer(),
      assembleDataCfg({
        meta: [],
        fields: {
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
        data: slice(originData, 0, 10),
      }),
      assembleOptions({
        seriesNumber: {
          enable: false,
        },
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: NewTab,
    });
    const rows = data.split(NewLine);
    const headers = rows[0].split(NewTab);

    // 33行数据 包括一行列头
    expect(rows).toHaveLength(11);
    // 5列数据 包括序列号
    rows.forEach((e) => {
      expect(e.split(NewTab)).toHaveLength(5);
    });
    expect(headers).toEqual(['province', 'city', 'type', 'sub_type', 'number']);
  });

  it('should export correct data with totals', async () => {
    const s2 = new TableSheet(
      getContainer(),
      assembleDataCfg({
        meta: [
          {
            field: 'province',
            formatter: (value) => `${value}-province`,
          },
          {
            field: 'type',
            formatter: (value) => `${value}-type`,
          },
        ],
        data: slice(originData, 0, 5),
        fields: {
          columns: ['province', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions({
        seriesNumber: {
          enable: false,
        },
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: NewTab,
      formatOptions: true,
    });

    expect(data).toMatchSnapshot();
  });

  it('should support custom export matrix transformer', async () => {
    const s2 = new TableSheet(
      getContainer(),
      assembleDataCfg({
        data: slice(originData, 0, 5),
        fields: {
          columns: ['province', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions({
        seriesNumber: {
          enable: false,
        },
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: NewTab,
      formatOptions: true,
      customTransformer: () => {
        return {
          [CopyMIMEType.PLAIN]: () => {
            return { type: CopyMIMEType.PLAIN, content: 'custom data' };
          },
        };
      },
    });

    expect(data).toMatchInlineSnapshot(`"custom data"`);
    expect(data).toEqual('custom data');
  });

  // https://github.com/antvis/S2/issues/2236
  it('should export correct data When the split separator is configured', async () => {
    const tableSheet = new TableSheet(
      getContainer(),
      assembleDataCfg({
        fields: {
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions(),
    );

    await tableSheet.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: tableSheet,
      split: ',',
    });
    // 只取前10行数据
    const result = slice(data.split(NewLine), 0, 5);

    expect(result).toMatchSnapshot();
  });

  it('should export correct data with formatter', async () => {
    const tableSheet = new TableSheet(
      getContainer(),
      assembleDataCfg({
        meta: [
          {
            field: 'number',
            name: '数量',
            description: '数量说明。。',
          },
          {
            field: 'province',
            name: '省份',
            description: '省份说明。。',
          },
          {
            field: 'city',
            name: '城市',
            description: '城市说明。。',
          },
          {
            field: 'type',
            name: '类别',
            description: '类别说明。。',
          },
          {
            field: 'sub_type',
            name: '子类别',
            description: '子类别说明。。',
          },
        ],
        fields: {
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions(),
    );

    await tableSheet.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: tableSheet,
      split: ',',
      formatOptions: {
        formatHeader: true,
      },
    });

    expect(data.split(NewLine)).toMatchSnapshot();
  });

  it('should export correct data with formatter if contain repeat column name', async () => {
    const tableSheet = new TableSheet(
      getContainer(),
      assembleDataCfg({
        meta: [
          {
            field: 'number',
            name: '子类别',
          },
          {
            field: 'province',
            name: '省份',
          },
          {
            field: 'city',
            name: '子类别',
          },
          {
            field: 'type',
            name: '类别',
          },
          {
            field: 'sub_type',
            name: '子类别',
          },
        ],
        fields: {
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions(),
    );

    await tableSheet.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: tableSheet,
      split: ',',
      formatOptions: {
        formatHeader: true,
      },
    });

    expect(data.split(NewLine)).toMatchSnapshot();
  });
});
