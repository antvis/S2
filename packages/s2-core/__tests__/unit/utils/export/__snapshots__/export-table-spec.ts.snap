// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableSheet Export Test should export correct data When the split separator is configured 1`] = `
Array [
  "province,city,type,sub_type,number",
  "浙江省,杭州市,家具,桌子,7789",
  "浙江省,绍兴市,家具,桌子,2367",
  "浙江省,宁波市,家具,桌子,3877",
  "浙江省,舟山市,家具,桌子,4342",
]
`;

exports[`TableSheet Export Test should export correct data with formatter 1`] = `
Array [
  "省份,城市,类别,子类别,数量",
  "浙江省,杭州市,家具,桌子,7789",
  "浙江省,绍兴市,家具,桌子,2367",
  "浙江省,宁波市,家具,桌子,3877",
  "浙江省,舟山市,家具,桌子,4342",
  "浙江省,杭州市,家具,沙发,5343",
  "浙江省,绍兴市,家具,沙发,632",
  "浙江省,宁波市,家具,沙发,7234",
  "浙江省,舟山市,家具,沙发,834",
  "浙江省,杭州市,办公用品,笔,945",
  "浙江省,绍兴市,办公用品,笔,1304",
  "浙江省,宁波市,办公用品,笔,1145",
  "浙江省,舟山市,办公用品,笔,1432",
  "浙江省,杭州市,办公用品,纸张,1343",
  "浙江省,绍兴市,办公用品,纸张,1354",
  "浙江省,宁波市,办公用品,纸张,1523",
  "浙江省,舟山市,办公用品,纸张,1634",
  "四川省,成都市,家具,桌子,1723",
  "四川省,绵阳市,家具,桌子,1822",
  "四川省,南充市,家具,桌子,1943",
  "四川省,乐山市,家具,桌子,2330",
  "四川省,成都市,家具,沙发,2451",
  "四川省,绵阳市,家具,沙发,2244",
  "四川省,南充市,家具,沙发,2333",
  "四川省,乐山市,家具,沙发,2445",
  "四川省,成都市,办公用品,笔,2335",
  "四川省,绵阳市,办公用品,笔,245",
  "四川省,南充市,办公用品,笔,2457",
  "四川省,乐山市,办公用品,笔,2458",
  "四川省,成都市,办公用品,纸张,4004",
  "四川省,绵阳市,办公用品,纸张,3077",
  "四川省,南充市,办公用品,纸张,3551",
  "四川省,乐山市,办公用品,纸张,352",
  ",,家具,桌子,26193",
  ",,家具,,49709",
  ",,家具,沙发,23516",
  ",,办公用品,,29159",
  ",,办公用品,笔,12321",
  ",,办公用品,纸张,16838",
  "浙江省,,家具,桌子,18375",
  "浙江省,,家具,沙发,14043",
  "浙江省,,办公用品,笔,4826",
  "浙江省,,办公用品,纸张,5854",
  "四川省,,家具,桌子,7818",
  "四川省,,家具,沙发,9473",
  "四川省,,办公用品,笔,7495",
  "四川省,,办公用品,纸张,10984",
  "浙江省,杭州市,家具,,13132",
  "浙江省,杭州市,办公用品,,2288",
  "浙江省,杭州市,,,15420",
  "浙江省,绍兴市,家具,,2999",
  "浙江省,绍兴市,办公用品,,2658",
  "浙江省,绍兴市,,,5657",
  "浙江省,宁波市,家具,,11111",
  "浙江省,宁波市,办公用品,,2668",
  "浙江省,宁波市,,,13779",
  "浙江省,舟山市,家具,,5176",
  "浙江省,舟山市,办公用品,,3066",
  "浙江省,舟山市,,,8242",
  "四川省,成都市,家具,,4174",
  "四川省,成都市,办公用品,,6339",
  "四川省,成都市,,,10513",
  "四川省,绵阳市,家具,,4066",
  "四川省,绵阳市,办公用品,,3322",
  "四川省,绵阳市,,,7388",
  "四川省,南充市,家具,,4276",
  "四川省,南充市,办公用品,,6008",
  "四川省,南充市,,,10284",
  "四川省,乐山市,家具,,4775",
  "四川省,乐山市,办公用品,,2810",
  "四川省,乐山市,,,7585",
  "浙江省,,家具,,32418",
  "浙江省,,办公用品,,10680",
  "浙江省,,,,43098",
  "四川省,,家具,,17291",
  "四川省,,办公用品,,18479",
  "四川省,,,,35770",
  ",,,,78868",
]
`;

exports[`TableSheet Export Test should export correct data with formatter if contain repeat column name 1`] = `
Array [
  "省份,子类别,类别,子类别,子类别",
  "浙江省,杭州市,家具,桌子,7789",
  "浙江省,绍兴市,家具,桌子,2367",
  "浙江省,宁波市,家具,桌子,3877",
  "浙江省,舟山市,家具,桌子,4342",
  "浙江省,杭州市,家具,沙发,5343",
  "浙江省,绍兴市,家具,沙发,632",
  "浙江省,宁波市,家具,沙发,7234",
  "浙江省,舟山市,家具,沙发,834",
  "浙江省,杭州市,办公用品,笔,945",
  "浙江省,绍兴市,办公用品,笔,1304",
  "浙江省,宁波市,办公用品,笔,1145",
  "浙江省,舟山市,办公用品,笔,1432",
  "浙江省,杭州市,办公用品,纸张,1343",
  "浙江省,绍兴市,办公用品,纸张,1354",
  "浙江省,宁波市,办公用品,纸张,1523",
  "浙江省,舟山市,办公用品,纸张,1634",
  "四川省,成都市,家具,桌子,1723",
  "四川省,绵阳市,家具,桌子,1822",
  "四川省,南充市,家具,桌子,1943",
  "四川省,乐山市,家具,桌子,2330",
  "四川省,成都市,家具,沙发,2451",
  "四川省,绵阳市,家具,沙发,2244",
  "四川省,南充市,家具,沙发,2333",
  "四川省,乐山市,家具,沙发,2445",
  "四川省,成都市,办公用品,笔,2335",
  "四川省,绵阳市,办公用品,笔,245",
  "四川省,南充市,办公用品,笔,2457",
  "四川省,乐山市,办公用品,笔,2458",
  "四川省,成都市,办公用品,纸张,4004",
  "四川省,绵阳市,办公用品,纸张,3077",
  "四川省,南充市,办公用品,纸张,3551",
  "四川省,乐山市,办公用品,纸张,352",
  ",,家具,桌子,26193",
  ",,家具,,49709",
  ",,家具,沙发,23516",
  ",,办公用品,,29159",
  ",,办公用品,笔,12321",
  ",,办公用品,纸张,16838",
  "浙江省,,家具,桌子,18375",
  "浙江省,,家具,沙发,14043",
  "浙江省,,办公用品,笔,4826",
  "浙江省,,办公用品,纸张,5854",
  "四川省,,家具,桌子,7818",
  "四川省,,家具,沙发,9473",
  "四川省,,办公用品,笔,7495",
  "四川省,,办公用品,纸张,10984",
  "浙江省,杭州市,家具,,13132",
  "浙江省,杭州市,办公用品,,2288",
  "浙江省,杭州市,,,15420",
  "浙江省,绍兴市,家具,,2999",
  "浙江省,绍兴市,办公用品,,2658",
  "浙江省,绍兴市,,,5657",
  "浙江省,宁波市,家具,,11111",
  "浙江省,宁波市,办公用品,,2668",
  "浙江省,宁波市,,,13779",
  "浙江省,舟山市,家具,,5176",
  "浙江省,舟山市,办公用品,,3066",
  "浙江省,舟山市,,,8242",
  "四川省,成都市,家具,,4174",
  "四川省,成都市,办公用品,,6339",
  "四川省,成都市,,,10513",
  "四川省,绵阳市,家具,,4066",
  "四川省,绵阳市,办公用品,,3322",
  "四川省,绵阳市,,,7388",
  "四川省,南充市,家具,,4276",
  "四川省,南充市,办公用品,,6008",
  "四川省,南充市,,,10284",
  "四川省,乐山市,家具,,4775",
  "四川省,乐山市,办公用品,,2810",
  "四川省,乐山市,,,7585",
  "浙江省,,家具,,32418",
  "浙江省,,办公用品,,10680",
  "浙江省,,,,43098",
  "四川省,,家具,,17291",
  "四川省,,办公用品,,18479",
  "四川省,,,,35770",
  ",,,,78868",
]
`;

exports[`TableSheet Export Test should export correct data with series number 1`] = `
Array [
  "序号	province	city	产品类型	sub_type	number",
  "1	浙江省	杭州市	家具	桌子	7789",
  "2	浙江省	绍兴市	家具	桌子	2367",
  "3	浙江省	宁波市	家具	桌子	3877",
  "4	浙江省	舟山市	家具	桌子	4342",
]
`;

exports[`TableSheet Export Test should export correct data with series number 2`] = `
Array [
  "序号	province	city	产品类型	sub_type	number",
  "1	浙江省	杭州市	家具	桌子	7789",
  "2	浙江省	绍兴市	家具	桌子	2367",
  "3	浙江省	宁波市	家具	桌子	3877",
  "4	浙江省	舟山市	家具	桌子	4342",
]
`;

exports[`TableSheet Export Test should export correct data with totals 1`] = `
"province	type	sub_type	number
浙江省-province	家具-type	桌子	7789
浙江省-province	家具-type	桌子	2367
浙江省-province	家具-type	桌子	3877
浙江省-province	家具-type	桌子	4342
浙江省-province	家具-type	沙发	5343"
`;
